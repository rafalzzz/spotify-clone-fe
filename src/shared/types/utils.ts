export type TCustomRequest = {
  basicUrl?: string;
  endpoint?: string;
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE';
  headers?: Record<string, string>;
  requestBody?: Record<string, unknown>;
  allowCookies?: boolean;
};
