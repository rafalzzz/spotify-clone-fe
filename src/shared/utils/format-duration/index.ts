const pad = (number: number): string => {
  return ('0' + number).slice(-2);
};

export const formatDuration = (seconds: number, isReversedTime: boolean = false): string => {
  if (seconds < 0) return '0:00';

  const date = new Date(seconds * 1000);
  const hh = date.getUTCHours();
  const mm = date.getUTCMinutes();
  const ss = pad(date.getUTCSeconds());

  let value = `${mm}:${ss}`;

  if (hh) {
    value = `${hh}:${pad(mm)}:${ss}`;
  }

  if (isReversedTime) {
    return `-${value}`;
  }

  return value;
};
