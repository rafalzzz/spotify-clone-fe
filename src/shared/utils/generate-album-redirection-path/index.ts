export const generateAlbumRedirectionPath = (artistName: string): string =>
  `/album/${artistName.toLocaleLowerCase()}`;
