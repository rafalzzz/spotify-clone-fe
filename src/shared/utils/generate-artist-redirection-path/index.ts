export const generateArtistRedirectionPath = (artistName: string): string =>
  `/artist/${artistName.toLocaleLowerCase()}`;
