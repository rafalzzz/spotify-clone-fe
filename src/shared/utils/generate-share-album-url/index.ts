export const generateShareAlbumUrl = (albumId: number): string =>
  window.location.origin + `/album/${albumId}`;
