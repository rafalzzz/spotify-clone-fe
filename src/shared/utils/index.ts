export { convertMusicTrackToSongItem } from './convert-music-track-to-song-item';
export { getItemsAmount } from './get-items-amount';
export { getMainArtist } from './get-main-artist';
export { generateAlbumRedirectionPath } from './generate-album-redirection-path';
export { generateArtistRedirectionPath } from './generate-artist-redirection-path';
export { generateShareTrackUrl } from './generate-share-track-url';
export { generateTrackRedirectionPath } from './generate-track-redirection-path';
export { getShuffledIndexes } from './get-shuffled-indexes';
