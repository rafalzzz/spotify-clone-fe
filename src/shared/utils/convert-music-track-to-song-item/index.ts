import { TSongItem } from '@/types/components';
import { EMusicTrackKeys, TMusicTrack } from '@/types/music-track';

export const convertMusicTrackToSongItem = (musicTrack: TMusicTrack): TSongItem => ({
  [EMusicTrackKeys.ARTIST_NAME]: musicTrack[EMusicTrackKeys.ARTIST_NAME],
  [EMusicTrackKeys.COLLECTION_NAME]: musicTrack[EMusicTrackKeys.COLLECTION_NAME],
  [EMusicTrackKeys.TRACK_NAME]: musicTrack[EMusicTrackKeys.TRACK_NAME],
  [EMusicTrackKeys.PREVIEW_URL]: musicTrack[EMusicTrackKeys.PREVIEW_URL],
  [EMusicTrackKeys.ARTWORK_URL_60]: musicTrack[EMusicTrackKeys.ARTWORK_URL_60],
  [EMusicTrackKeys.ARTIST_ID]: musicTrack[EMusicTrackKeys.ARTIST_ID],
  [EMusicTrackKeys.COLLECTION_ID]: musicTrack[EMusicTrackKeys.COLLECTION_ID],
  [EMusicTrackKeys.TRACK_ID]: musicTrack[EMusicTrackKeys.TRACK_ID],
});
