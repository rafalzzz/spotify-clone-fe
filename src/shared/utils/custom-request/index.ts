import { TCustomRequest } from '@/types/utils';

export const customRequest = ({
  basicUrl = process.env.NEXT_PUBLIC_API_URL,
  endpoint = '/',
  method,
  headers = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  },
  requestBody,
  allowCookies = false,
}: TCustomRequest): Promise<Response> =>
  fetch(basicUrl + endpoint, {
    method,
    headers,
    body: JSON.stringify(requestBody),
    credentials: allowCookies ? 'include' : 'omit',
  });
