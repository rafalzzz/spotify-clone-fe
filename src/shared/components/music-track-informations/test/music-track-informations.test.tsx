import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import { generateArtistRedirectionPath } from '@/utils/generate-artist-redirection-path';

import { MusicTrackInformations } from '../';

const MOCKED_TRACK_NAME = 'Mocked title';

const FIRST_ARTIST_NAME = 'Mocked First Artist Name';
const MOCKED_ARTIST_NAME = `${FIRST_ARTIST_NAME}, Second Artist Name`;

const TRACK_NAME_TEST_ID = 'music-track-informations-track-name';
const ARTIST_NAME_TEST_ID = 'music-track-informations-artist-name';
const TRACK_NAME_TOOLTIP_TEST_ID = 'music-track-informations-track-name-tooltip';
const ARTIST_NAME_TOOLTIP_TEST_ID = 'music-track-informations-artist-name-tooltip';

const renderMusicTrackInformations = () =>
  render(<MusicTrackInformations trackName={MOCKED_TRACK_NAME} artistName={MOCKED_ARTIST_NAME} />);

describe('MusicTrackInformations', () => {
  it('renders component without error', () => {
    const screen = renderMusicTrackInformations();
    expect(screen).toMatchSnapshot();
  });

  it('renders the track name and main artist name', () => {
    const { queryByTestId } = renderMusicTrackInformations();

    const trackNameElement = queryByTestId(TRACK_NAME_TEST_ID);
    const artistNameElement = queryByTestId(ARTIST_NAME_TEST_ID);

    expect(trackNameElement).toBeInTheDocument();
    expect(artistNameElement).toBeInTheDocument();
  });

  it('displays only the main artist when multiple artists are given', () => {
    const { queryByTestId } = renderMusicTrackInformations();

    const artistNameElement = queryByTestId(ARTIST_NAME_TEST_ID);
    expect(artistNameElement).toHaveTextContent(FIRST_ARTIST_NAME);
  });

  it('passes track name to the track tooltip', () => {
    const { queryByTestId, queryByText } = renderMusicTrackInformations();

    const trackNameTooltip = queryByTestId(TRACK_NAME_TOOLTIP_TEST_ID);
    expect(trackNameTooltip).toBeInTheDocument();

    const tooltipTitle = queryByText(MOCKED_TRACK_NAME, { selector: 'span' });
    expect(tooltipTitle).toBeInTheDocument();
  });

  it('passes main artist name to the artist tooltip', () => {
    const { queryByTestId } = renderMusicTrackInformations();

    const artistNameTooltip = queryByTestId(ARTIST_NAME_TOOLTIP_TEST_ID);
    expect(artistNameTooltip).toBeInTheDocument();

    const tooltipTitle = (queryByTestId(ARTIST_NAME_TEST_ID) as HTMLElement).querySelector('span');
    expect(tooltipTitle).toBeInTheDocument();
  });

  it('artist redirection have correct path', () => {
    const { queryByTestId } = renderMusicTrackInformations();

    const artistNameElement = queryByTestId(ARTIST_NAME_TEST_ID);
    expect(artistNameElement).toHaveAttribute(
      'href',
      generateArtistRedirectionPath(FIRST_ARTIST_NAME),
    );
  });
});
