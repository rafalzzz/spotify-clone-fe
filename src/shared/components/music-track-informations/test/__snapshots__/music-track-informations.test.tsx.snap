// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MusicTrackInformation renders component without error 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="music-track-informations"
      >
        <div
          data-mouseenterdelay="0.5"
          data-placement="right"
          data-testid="music-track-informations-track-name-tooltip"
        >
          <h4
            class="music-track-informations__text"
            data-testid="music-track-informations-track-name"
          >
            Mocked title
          </h4>
          <span>
            Mocked title
          </span>
        </div>
        <button
          class="music-track-informations__text music-track-informations__artist"
          data-testid="music-track-informations-artist-name"
        >
          <div
            data-mouseenterdelay="0.5"
            data-placement="right"
            data-testid="music-track-informations-artist-name-tooltip"
          >
            <span>
              Mocked First TArtist Name
            </span>
            <span>
              Mocked First TArtist Name
            </span>
          </div>
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="music-track-informations"
    >
      <div
        data-mouseenterdelay="0.5"
        data-placement="right"
        data-testid="music-track-informations-track-name-tooltip"
      >
        <h4
          class="music-track-informations__text"
          data-testid="music-track-informations-track-name"
        >
          Mocked title
        </h4>
        <span>
          Mocked title
        </span>
      </div>
      <button
        class="music-track-informations__text music-track-informations__artist"
        data-testid="music-track-informations-artist-name"
      >
        <div
          data-mouseenterdelay="0.5"
          data-placement="right"
          data-testid="music-track-informations-artist-name-tooltip"
        >
          <span>
            Mocked First TArtist Name
          </span>
          <span>
            Mocked First TArtist Name
          </span>
        </div>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
